internal/invocation/invocation.go

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package invocation
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	pubcluster "github.com/hazelcast/hazelcast-go-client/cluster"
+	"github.com/hazelcast/hazelcast-go-client/internal/hzerror"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto"
+)
+
+type Result interface {
+	Get() (*proto.ClientMessage, error)
+	GetWithTimeout(duration time.Duration) (*proto.ClientMessage, error)
+}
+
+type Invocation interface {
+	Complete(message *proto.ClientMessage)
+	Completed() bool
+	EventHandler() proto.ClientMessageHandler
+	Get() (*proto.ClientMessage, error)
+	GetContext(ctx context.Context) (*proto.ClientMessage, error)
+	GetWithTimeout(duration time.Duration) (*proto.ClientMessage, error)
+	PartitionID() int32
+	Request() *proto.ClientMessage
+	Address() *pubcluster.AddressImpl
+}
+
+type Impl struct {
+	request         *proto.ClientMessage
+	response        chan *proto.ClientMessage
+	completed       int32
+	boundConnection *Impl
+	address         *pubcluster.AddressImpl
+	partitionID     int32
+	//sentConnection  atomic.Value
+	eventHandler func(clientMessage *proto.ClientMessage)
+	deadline     time.Time
```

Comment:
I could not find a usage of `deadline`. Is it missing?

Reply:
Implemented at: https://github.com/yuce/hazelcast-go-client/commit/65e58171e8b415d9f55907da4de1ed46abcb0238

Reply:
I'm not sure if changes in 65e58171e8b415d9f55907da4de1ed46abcb0238 are enough for a proper implementation. The thing is that invocation has to be deregistered from the invocation service once the deadline is reached. Java client has `InvocationMonitor` responsible for scanning through pending invocations, detecting deadlines for them (incl. backup timeouts, but that's for boomerang backup feature) and de-registering them. Go client may need to implement something similar, otherwise stale invocations will lead to mem leaks.

See: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java#L304

Reply:
Sorry for pointing you at member-side invocation code. The proper link should be this connection closed handler: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientInvocationServiceImpl.java#L217